src/org/zaproxy/zap/extension/ascanrulesAlpha/HttpOnlySite.java

```diff
@@ -156,58 +152,70 @@ public void scan() {
+			String scheme = "http"; //If any redirection is https scheme will be changed to https
 			int count = 0;
 			while(count < 10) {
+				if (isStop()) {
+					if (log.isDebugEnabled()) {
+						log.debug("Scanner "+getName()+" Stopping.");
+					}
+					return;
+				}
 				sendAndReceive(newRequest, false);
-				System.out.println("\n\nThe request was succesful and it was\n"+newRequest.getRequestHeader().toString());
-				System.out.println("\n\nA response was recieved and it was\n"+newRequest.getResponseHeader().toString());
-				System.out.println("\n\nThe status code is: "+newRequest.getResponseHeader().getStatusCode());
-				String loc = newRequest.getResponseHeader().LOCATION;
-				System.out.println("\n\nThe location is: "+newRequest.getResponseHeader().getHeader(loc));
 				int status = newRequest.getResponseHeader().getStatusCode();
-				String redirect = newRequest.getResponseHeader().getHeader(loc);
-				if(status == HttpStatusCode.MOVED_PERMANENTLY || status == HttpStatusCode.FOUND) {
-					String path = newRequest.getRequestHeader().getURI().getPath();
-					if(redirect.contains("https")) {
-						String newHost = redirect.substring(8, redirect.length());
-					    newRequest.getRequestHeader().setURI(new URI("https", null, newHost, 443, path));
-					    record = "https";
-					    count++;
-					} else if (redirect.contains("http")) {
-						raiseAlert(newRequest);
-						break;
+				if(HttpStatusCode.isRedirection(status)) {
+					String loc = HttpResponseHeader.LOCATION;
+					String redirect = newRequest.getResponseHeader().getHeader(loc);
+					if(!redirect.isEmpty() && redirect != null) {
+						URI oldURI = newRequest.getRequestHeader().getURI();
+						try {
+							newRequest.getRequestHeader().setURI(new URI(oldURI, redirect, true));
+							URI newURI = newRequest.getRequestHeader().getURI();
+							if(!oldURI.getHost().equals(newURI.getHost())) {
+								String message = "Different hosts";
+								raiseAlert(newRequest, message);
+								return;
+							}
+						} catch (URIException e) {
+							try {
+								newRequest.getRequestHeader().setURI(new URI(oldURI, redirect, false));
+							} catch (URIException ex) {
+								String message = "URL could not be encoded";
+								raiseAlert(newRequest, message);
+								return;
+							}
+						}
+						if(newRequest.getRequestHeader().getURI().getScheme().equals("https")) {
+						    scheme = "https";
+						    count++;
+						} else if (redirect.startsWith("http")) {
+							String message = "Redirection was to an HTTP page";
+							raiseAlert(newRequest, message);
+							return;
+						}	
 					}
-				}
-				else {
+				}else {
 					break;
 				}
 			}
-			if(record.equals("http") && count == 10) { //For cases when even the 10th redirection is not https
-				raiseAlert(newRequest);
+			if(scheme.equals("http") && count == 10) { //For cases when even the 10th redirection is not https
+				String message = "Redirection was to an HTTP page";
```

Comment:
Should also tell that was after following 10 redirections (i.e. the maximum was reached).